src/ir/ir_printer.h

```diff
@@ -0,0 +1,133 @@
+//-----------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+#ifndef SRC_IR_IR_PRINTER_H_
+#define SRC_IR_IR_PRINTER_H_
+
+#include <memory>
+
+#include "absl/strings/str_format.h"
+#include "absl/strings/str_join.h"
+#include "src/ir/ir_traversing_visitor.h"
+#include "src/ir/module.h"
+#include "src/ir/value.h"
+
+namespace raksha::ir {
+
+// A visitor that prints the IR.
+class IRPrinter : public IRTraversingVisitor<IRPrinter> {
+ public:
+  template <typename T>
+  static std::string ToString(const T& entity) {
+    IRPrinter printer;
+    entity.Accept(printer);
+    return printer.out.str();
+  }
```

Comment:
I think, in the common case that this is printed to a log file or the console, it will be friendlier to have an interface like

```C++
static void ToString(const T& entity, ostream &out)
```

where `out` is used to construct the `IRPrinter` which ends up holding an `ostream &` field. That will allow everything to be printed out directly to the console/file rather than buffering in an `ostringstream` first. You can still have an alternative with the above interface by just having an implementation like this:

```C++
static std::string ToString(const T& entity) {
  std::ostringstream out;
  ToString(entity, out);
  return out.str;
}
```

Reply:
That was what I had in mind originally, but I did not like having to hold a reference to `ostream` because the constructor was public at that time. Now that the constructor is private, I think it is fine to do so. Will do.

Reply:
done.